influxdb3_write/src/last_cache.rs

```diff
@@ -0,0 +1,2225 @@
+    ///
+    /// This will default to [`DEFAULT_CACHE_TTL`]
+    pub(crate) ttl: Option<Duration>,
+    /// The key column names to use in the cache hierarchy
+    ///
+    /// This will default to:
+    /// - the series key columns for a v3 table
+    /// - the lexicographically ordered tag set for a v1 table
+    pub(crate) key_columns: Option<Vec<String>>,
+    /// The value columns to use in the cache
+    ///
+    /// This will default to all non-key columns. The `time` column is always included.
+    pub(crate) value_columns: Option<Vec<String>>,
+}
+
+impl LastCacheProvider {
+    /// Create a new [`LastCacheProvider`]
+    pub(crate) fn new() -> Self {
+        Self {
+            cache_map: Default::default(),
+        }
+    }
+
+    /// Create a new entry in the last cache for a given database and table, along with the given
+    /// parameters.
+    pub(crate) fn create_cache(
+        &self,
+        CreateCacheArguments {
+            db_name,
+            tbl_name,
+            schema,
+            cache_name,
+            count,
+            ttl,
+            key_columns,
+            value_columns,
+        }: CreateCacheArguments,
+    ) -> Result<(), Error> {
+        let key_columns = if let Some(keys) = key_columns {
+            // validate the key columns specified to ensure correct type (string, int, unit, or bool)
+            // and that they exist in the table's schema.
+            for key in keys.iter() {
+                use InfluxColumnType::*;
+                use InfluxFieldType::*;
+                match schema.field_by_name(key) {
+                    Some((
+                        Tag | Field(Integer) | Field(UInteger) | Field(String) | Field(Boolean),
+                        _,
+                    )) => (),
+                    Some((_, _)) => return Err(Error::InvalidKeyColumn),
+                    None => {
+                        return Err(Error::KeyColumnDoesNotExist {
+                            column_name: key.into(),
+                        })
+                    }
+                }
+            }
+            keys
+        } else {
+            // use primary key, which defaults to series key if present, then lexicographically
+            // ordered tags otherwise, there is no user-defined sort order in the schema, so if that
+            // is introduced, we will need to make sure that is accommodated here.
+            let mut keys = schema.primary_key();
+            if let Some(&TIME_COLUMN_NAME) = keys.last() {
+                keys.pop();
+            }
+            keys.iter().map(|s| s.to_string()).collect()
+        };
+
+        // Generate the cache name if it was not provided
+        let cache_name = cache_name.unwrap_or_else(|| {
+            format!("{tbl_name}_{keys}_last_cache", keys = key_columns.join("_"))
+        });
+
+        // reject creation if there is already a cache with specified database, table, and cache name
```

Comment:
The behavior I like to see on creation of things is that if the user tries to create something with all the same arguments, it returns ok, if the arguments are different, then it returns with an error that it already exists. This might be handled higher up the stack in the API layer, but I'm not sure that this error would give the caller enough information to make the determination.

The reason I opt for this behavior is that it makes automation easier where they can call create as many times as they want and it'll always succeed as long as the settings are the same.